source/agora/network/NetworkClient.d

```diff
@@ -208,18 +179,73 @@ class NetworkClient
 
             If the request failed, returns an empty array
 
+        Throws:
+            RequestFailure if the request fails
+
     ***************************************************************************/
 
     public Block[] getBlocksFrom (ulong block_height, size_t max_blocks)
     {
-        try
-        {
-            return this.api.getBlocksFrom(block_height, max_blocks);
-        }
-        catch (Exception ex)
+        return this.attemptRequest(
+            this.api.getBlocksFrom(block_height, max_blocks),
+            this.request_exc);
+    }
+
+    /***************************************************************************
+
+        Attempt a request up to MaxRetries attempts, and make the task
+        wait this.retry_delay between each attempt.
+
+        If all requests fail and 'ex' is not null, throw the exception.
+
+        Params:
+            dg = the delegate to call
+            ex = the exception to throw if all attempts fail (if not null)
+
+        Returns:
+            the return value of dg(), which may be void
+
+    ***************************************************************************/
+
+    private T attemptRequest (T) ( lazy T dg, Exception ex = null)
+    {
+        foreach (_; 0 .. MaxRetries)
         {
-            logError(ex.message);
-            return null;
+            try
+            {
+                return dg();
+            }
+            catch (Exception ex)
+            {
+                this.banman.onFailedRequest(this.address);
+                logError(ex.message);
+                this.taskman.wait(this.retry_delay);
+            }
         }
+
+        if (ex !is null)
+            throw ex;
+
+        // ex is null, failure is ignored
+        static if (!is(T == void))
+            return T.init;
+    }
+}
+
+/// Thrown when a request fails after a number of retries
+public class RequestFailure : Exception
```

Comment:
Question: continue using this or use an Exception?

I would like to avoid programming errors to throw exceptions and be caught as if they were request exceptions. 

For example, if I have a call like `string stuff = format("%s %s", 1);`, this will throw an Exception. But it's really a programmer bug, not a "failed request". See what I mean?